### ISSUE TITLE
If the first poke of a sensor throws an exception, `timeout` does not work

### ISSUE THREAD

**Issue by chrisranderson** [apache/airflow#30289]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Airflow Version: 2.2.5

In `reschedule` mode, if the first poke of a sensor throws an exception, `timeout` does not work. There can be any combination of the poke returning `False` or raising an exception after that. My guess is that something is initialized in some database incorrectly, because this returns an empty list every time if the first poke raises an exception:

```
TaskReschedule.find_for_task_instance(
            context["ti"], try_number=first_try_number
        )
```

This happens here in the main branch: https://github.com/apache/airflow/blob/main/airflow/sensors/base.py#L174-L181

If the first poke returns `False`, I don't see this issue. 

### What you think should happen instead

The timeout should be respected whether `poke` returns successfully or not. 

A related issue is that if every poke raises an uncaught exception, the timeout will never be respected, since the timeout is checked only after a successful poke. Maybe both issues can be fixed at once?

### How to reproduce

Use this code. Run the dag several times, and see if the total duration including all retries is greater than the timeout.

```
import datetime
import random

from airflow import DAG
from airflow.models import TaskReschedule
from airflow.sensors.base import BaseSensorOperator
from airflow.utils.context import Context


class RandomlyFailSensor(BaseSensorOperator):
    def poke(self, context: Context) -> bool:
        first_try_number = context["ti"].max_tries - self.retries + 1
        task_reschedules = TaskReschedule.find_for_task_instance(
            context["ti"], try_number=first_try_number
        )
        self.log.error(f"\n\nIf this is the first attempt, or first attempt failed, "
                       f"this will be empty: \n\t{task_reschedules}\n\n")

        if random.random() < .5:
            self.log.error("\n\nIf this was the very first poke, the timeout *will not* work.\n\n")
            raise Exception('Failed!')
        else:
            self.log.error("\n\nIf this was the very first poke, the timeout *will* work.\n\n")
            return False


dag = DAG(
    'sensors_test',
    schedule_interval=None,
    max_active_runs=1,
    catchup=False,
    default_args={
        "owner": "me",
        "depends_on_past": False,
        "start_date": datetime.datetime(2018, 1, 1),
        "email_on_failure": False,
        "email_on_retry": False,
        "execution_timeout": datetime.timedelta(minutes=10),
    }
)

t_always_fail_sensor = RandomlyFailSensor(
    task_id='random_fail_sensor',
    mode="reschedule",
    poke_interval=1,
    retry_delay=datetime.timedelta(seconds=1),
    timeout=15,
    retries=50,
    dag=dag
)
```

### Operating System

Debian 11? This Docker image: https://hub.docker.com/layers/library/python/3.8.12/images/sha256-60d1cda1542582095795c25bff869b0c615e2a913c4026ed0313ede156b60468?context=explore

### Deployment

Docker-Compose

### Deployment details

I use an internal tool that hides the details of the deployment. If there is more info that would be helpful for debugging, let me know.

### Anything else

- This happens every time, based on the conditions I describe above.
- I'd be happy to submit a PR, but that depends on what my manager says.
- @yuqian90 might know more about this issue, since they contributed related code in [this commit](https://github.com/apache/airflow/commit/a0e6a847aa72ddb15bdc147695273fb3aec8839d#diff-62f7d8a52fefdb8e05d4f040c6d3459b4a56fe46976c24f68843dbaeb5a98487R1164).
- Impact: if the first poke throws an exception and the rest return False, the task will continue indefinitely.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
I am not sure if we should consider this as a bug or a new feature. (@potiuk)

When the poke method raises an exception, the sensor task fails, and a new task instance will be created if you provide `retries` > 0 (50 in your example).

For now, you should handle this failure by yourself, and return `False` when you want to skip the exception. But as a new feature, I added in #30293 a new param `silent_fail` to catch the exceptions (except 4 Airflow exceptions) and log it instead of failing the task, in this case the execution will be continue in the same retry.

WDYT?

**Comment by chrisranderson** [comment]
> For now, you should handle this failure by yourself, and return False when you want to skip the exception.

We initially handled it by subclassing BaseSensorOperator and implementing `poke` with a big try-except around the whole function. I'd like to avoid this since we would have to fix every built-in sensor that we use by wrapping all their poke functions.  We'd also have to adjust retries, since this makes it a lot harder to trigger a retry. Before, they would be triggered by any exception. Now it would only be ones we specify or something like the worker shutting down.

`silent_fail` doesn't schedule a new attempt - I would hope a new attempt would start, *and* the timeout is respected. I want to say "retry 3 times" and have retries cover any uncaught exception, not just extreme ones like the worker being shut down.

**Comment by chrisranderson** [comment]
> check the timeout before the poke

This won't totally solve the issue, though - even with this, if the first poke fails, and it never fails again, it will continue forever. It won't retry because it doesn't fail, and it won't hit the timeout because something wasn't initialized correctly and future timeout checks (i.e. `TaskReschedule.find_for_task_instance`) won't work.

**Comment by hussein-awala** [comment]
> That could work, but it's probably out of the scope of this issue? But maybe that's a nice enhancement for operators that aren't sensors. I don't know.

The sensors are running as task instances (`class BaseSensorOperator(BaseOperator, SkipMixin)`), so if we implement this, we can timeout the ti regardless the number of retries and the result of each retry. Why do you think it is out of the scope of this issue?

Since some users expect to reset `timeout` and `execution_timeout` after each retry, I think we should add a new conf to implement the solution whatever this solution is.

**Comment by potiuk** [comment]
I agree the solution by @hussein-awala is really good approach. The "unhandled exception" from `poke()` is not really "expected" and thus the behaviour is really undefined. Adding the new parameter handles it nicely if you want to do it, while keeping backwards compatibile behaviour.
